sdk/core/core-https/test/challengeAuthenticationPolicy.spec.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { assert } from "chai";
+import { AccessToken, GetTokenOptions, TokenCredential } from "@azure/core-auth";
+import {
+  bearerTokenAuthenticationPolicy,
+  challengeAuthenticationPolicy,
+  createEmptyPipeline,
+  createHttpHeaders,
+  createPipelineRequest,
+  HttpsClient,
+  PipelineRequest,
+  PipelineResponse
+} from "../src";
+import { parseCAEChallenges } from "../src/util/parseCAEChallenges";
+
+export interface TestChallenge {
+  scope: string;
+  claims: string;
+}
+
+let cachedChallenge: string | undefined;
+
+/**
+ * Converts a uint8Array to a string.
+ */
+export function uint8ArrayToString(ab: Uint8Array): string {
+  const decoder = new TextDecoder("utf-8");
+  return decoder.decode(ab);
+}
+
+/**
+ * Encodes a string in base64 format.
+ * @param value - The string to encode
+ */
+export function encodeString(value: string): string {
+  return Buffer.from(value).toString("base64");
+}
+
+/**
+ * Decodes a base64 string into a byte array.
+ * @param value - The base64 string to decode
+ */
+export function decodeString(value: string): Uint8Array {
+  return Buffer.from(value, "base64");
+}
+
+async function processChallenge(request: PipelineRequest, challenge?: string): Promise<boolean> {
+  if (!challenge) {
+    return false;
+  }
+
+  const challenges: TestChallenge[] = parseCAEChallenges(challenge) || [];
+
+  const parsedChallenge = challenges.find((x) => x.claims);
+  if (!parsedChallenge) {
+    throw new Error("Missing claims");
+  }
+  if (cachedChallenge !== challenge) {
+    cachedChallenge = challenge;
+  }
+
+  if (!request.authenticationOptions) {
+    request.authenticationOptions = {};
+  }
+
+  request.authenticationOptions.scope = parsedChallenge.scope;
+  request.authenticationOptions.claims = uint8ArrayToString(decodeString(parsedChallenge.claims));
```

Comment:
This is one of those important design decisions that I'm not convinced was able to capture properly. The main idea is that we need a way to set the scope of the next getToken request, and the claims that will be sent as part of the authentication too.

C# uses a "getTokenRequestContext", an object that they pass to this processChallenge function that is used to store this scope and claims. I used a new plain object property in the PipelineRequest.

We could also use the return type, something like `processChallenge(challenge): { scope, claims } | null`, where if it's null, we don't proceed with the next request, and if it's not null, we alter the next getToken request with the received values.

What sounds like the better approach?